packages/wdio-browserstack-service/src/launcher.ts

```diff
@@ -291,20 +333,124 @@ export default class BrowserstackLauncherService implements Services.ServiceInst
+            this._updateCaps(capabilities, 'buildIdentifier')
+            return
+        }
+
+        if (!this._buildName || process.env.BROWSERSTACK_BUILD_NAME) {
+            if (this._buildIdentifier) {
+                this._updateCaps(capabilities, 'buildIdentifier')
+                log.warn('Skipping buildIdentifier as buildName is not passed.')
+                return
+            }
+        }
+
+        if (this._buildIdentifier && this._buildIdentifier.includes('${DATE_TIME}')){
+            const formattedDate = new Intl.DateTimeFormat('en-GB', {
+                month: 'short',
+                day: '2-digit',
+                hour: '2-digit',
+                minute: '2-digit',
+                hour12: false })
+                .format(new Date())
+                .replace(/ |, /g, '-')
+            this._buildIdentifier = this._buildIdentifier.replace('${DATE_TIME}', formattedDate)
+            this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
+        }
+
+        if (!this._buildIdentifier.includes('${BUILD_NUMBER}')) return
+
+        const ciInfo = getCiInfo()
+        if (ciInfo != null) {
+            const ciBuildNumber = ciInfo.build_number
+            if (ciBuildNumber != null) {
+                this._buildIdentifier = this._buildIdentifier.replace('${BUILD_NUMBER}', 'CI '+ ciBuildNumber)
+                this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
+            }
+        } else {
+            const localBuildNumber = this._getLocalBuildNumber()
+            if (localBuildNumber != '-1') {
+                this._buildIdentifier = this._buildIdentifier.replace('${BUILD_NUMBER}', localBuildNumber)
+                this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
+            } else {
+                return
+            }
+        }
+    }
+
+    _getLocalBuildNumber() {
+        let browserstackFolderPath = path.join(require('os').homedir(), '.browserstack')
+        try {
+            if (!fs.existsSync(browserstackFolderPath)){
+                fs.mkdirSync(browserstackFolderPath)
+            }
+
+            let filePath = path.join(browserstackFolderPath, '.build-name-cache.json')
+            if (!fs.existsSync(filePath)) {
+                fs.appendFileSync(filePath, JSON.stringify({}))
+            }
+
+            const buildCacheFileData = fs.readFileSync(filePath)
+            const parsedBuildCacheFileData = JSON.parse(buildCacheFileData.toString())
+
+            if (this._buildName && this._buildName in parsedBuildCacheFileData) {
+                let prevIdentifier = parseInt((parsedBuildCacheFileData[this._buildName]['identifier']))
+                let newIdentifier = prevIdentifier + 1
+                this._updateLocalBuildCache(filePath, this._buildName, newIdentifier)
+                return newIdentifier.toString()
+            }
+            this._updateLocalBuildCache(filePath, this._buildName, 1)
+            return '1'
```

Comment:
Make this more readable.
```suggestion
            const newIdentifier = 1
            this._updateLocalBuildCache(filePath, this._buildName, newIdentifier)
            return newIdentifier.toString()
```

Reply:
Incorporated all suggestions. Thank you Christian.